src/HTTP.jl

```diff
@@ -456,16 +460,18 @@ hard-codes the value of the `retry` and `redirect` keyword arguments. When we pa
 effectively over-writing the default and any user-provided keyword arguments for `retry` or `redirect`.
 """
 macro client(requestlayers, streamlayers=[])
-    esc(quote
-        get(a...; kw...) = request("GET", a...; kw...)
-        put(a...; kw...) = request("PUT", a...; kw...)
-        post(a...; kw...) = request("POST", a...; kw...)
-        patch(a...; kw...) = request("PATCH", a...; kw...)
-        head(a...; kw...) = request("HEAD", a...; kw...)
-        delete(a...; kw...) = request("DELETE", a...; kw...)
-        open(f, a...; kw...) = request(a...; iofunction=f, kw...)
-        request(method, url, h=HTTP.Header[], b=HTTP.nobody; headers=h, body=b, query=nothing, kw...)::HTTP.Response =
+    return @remove_linenums! esc(quote
+        get(a...; kw...) = ($__source__; request("GET", a...; kw...))
+        put(a...; kw...) = ($__source__; request("PUT", a...; kw...))
+        post(a...; kw...) = ($__source__; request("POST", a...; kw...))
+        patch(a...; kw...) = ($__source__; request("PATCH", a...; kw...))
+        head(a...; kw...) = ($__source__; request("HEAD", a...; kw...))
+        delete(a...; kw...) = ($__source__; request("DELETE", a...; kw...))
+        open(f, a...; kw...) = ($__source__; request(a...; iofunction=f, kw...))
+        function request(method, url, h=HTTP.Header[], b=HTTP.nobody; headers=h, body=b, query=nothing, kw...)::HTTP.Response
+            $__source__
             HTTP.request(HTTP.stack($requestlayers, $streamlayers), method, url, headers, body, query; kw...)
+        end
     end)
```

Comment:
```suggestion
    return Base.remove_linenums(quote
        get(a...; kw...) = ($__source__; request("GET", a...; kw...))
        put(a...; kw...) = ($__source__; request("PUT", a...; kw...))
        post(a...; kw...) = ($__source__; request("POST", a...; kw...))
        patch(a...; kw...) = ($__source__; request("PATCH", a...; kw...))
        head(a...; kw...) = ($__source__; request("HEAD", a...; kw...))
        delete(a...; kw...) = ($__source__; request("DELETE", a...; kw...))
        open(f, a...; kw...) = ($__source__; request(a...; iofunction=f, kw...))
        function request(method, url, h=HTTP.Header[], b=HTTP.nobody; headers=h, body=b, query=nothing, kw...)::HTTP.Response
            $__source__
            HTTP.request(HTTP.stack($(esc(requestlayers)), $(esc(streamlayers)), method, url, headers, body, query; kw...)
        end
    end)
```

Reply:
if doing more fine-grained `esc`-ing, i think we'd also need something like `$(esc(:get))(a...; kw...) = ...` so that the `get` is a new function (not gensym'd and not overloading `HTTP.get` nor `Base.get`) but 👍 i still think your proposal in https://github.com/JuliaLang/julia/pull/31335 is sound

Reply:
Ah yes, you're right. 👍 
For some reason i thought I remembered there was an exception in the macroexpansion rules that they _don't_ mangle the names of top-level function defs in a quote expression... but it seems you are correct:
```julia
julia> macro foo(value)
           return quote
               new_func() = $(esc(value))
           end
       end
@foo (macro with 1 method)

julia> @foo 2+2
#16#new_func (generic function with 1 method)

julia> new_func()
ERROR: UndefVarError: new_func not defined
```